annotator/src/test/java/org/cbioportal/annotator/internal/GenomeNexusImplTest.java

```diff
@@ -0,0 +1,331 @@
+    }
+
+    private final String ISOFORM_OVERRIDE_SOURCE = "uniprot";
+    private final Logger log = Logger.getLogger(GenomeNexusImplTest.class);
+
+    /**
+     * Test resolved variant classifications (mutation types) returned by Genome Nexus.
+     */
+    @Test
+    public void testResolveVariantClassification() {
+        StringBuilder errorMessage = new StringBuilder("\nFailures:\n");
+        int failCount = 0;
+
+        Map<String, String> expectedMutationTypes = makeMockExpectedMutationType();
+        for (AnnotatedRecord record : mockAnnotatedRecords) {
+            if (!record.getVARIANT_CLASSIFICATION().equals(expectedMutationTypes.get(record.getTUMOR_SAMPLE_BARCODE()))) {
+                errorMessage.append("testResolveVariantClassification(), annotated record variant classification '")
+                        .append(record.getVARIANT_CLASSIFICATION())
+                        .append("' does not match expected type '")
+                        .append(expectedMutationTypes.get(record.getTUMOR_SAMPLE_BARCODE()))
+                        .append(" for record '")
+                        .append(record.getTUMOR_SAMPLE_BARCODE())
+                        .append("'\n");
+                failCount += 1;
+            }
+        }
+
+        if (failCount > 0) {
+            Assert.fail(errorMessage.toString());
+        }
+    }
+
+    /**
+     * Test resolved protein changes returned by Genome Nexus.
+     */
+    @Test
+    public void testResolveProteinChange() {
+        StringBuilder errorMessage = new StringBuilder("\nFailures:\n");
+        int failCount = 0;
+
+        Map<String, String> expectedProteinChanges = makeMockExpectedProteinChange();
+        for (AnnotatedRecord record : makeMockAnnotatedRecords()) {
+            if (!record.getHGVSP_SHORT().equals(expectedProteinChanges.get(record.getTUMOR_SAMPLE_BARCODE()))) {
+                errorMessage.append("testResolveProteinChange(), annotated record protein change '")
+                        .append(record.getHGVSP_SHORT())
+                        .append("' does not match expected protein change '")
+                        .append(expectedProteinChanges.get(record.getTUMOR_SAMPLE_BARCODE()))
+                        .append(" for record '")
+                        .append(record.getTUMOR_SAMPLE_BARCODE())
+                        .append("'\n");
+                failCount += 1;
+            }
+        }
+
+        if (failCount > 0) {
+            Assert.fail(errorMessage.toString());
+        }
+    }
+
+    private List<AnnotatedRecord> makeMockAnnotatedRecords() {
+        List<MutationRecord> mockMutationRecords = makeMockMutationRecords();
+
+        List<AnnotatedRecord> mockAnnotatedRecords = new ArrayList();
+        for (MutationRecord record : mockMutationRecords) {
+            AnnotatedRecord annotatedRecord = null;
+            try {
+                annotatedRecord = annotator.annotateRecord(record, true, ISOFORM_OVERRIDE_SOURCE, true);
```

Comment:
This seems like an actual call to Genome Nexus to annotate records -- for a unit test, I generally think that we want to isolate the tested code from any dependencies. Whether Genome Nexus is available or not, or a change to the genome nexus annotation behavior probably should not cause unit tests to fail here. I'm not as familiar with how the annotator codebase is organized, but I think I would recommend either (1) encapsulating the functionality which sends requests to the genome nexus web service into its own class (a @Repository?) and then mocking that class to return pre-computed annotated records to support this unit test, or (2) isolating the business logic that transforms the genome nexus annotation into AnnotatedRecords (a @Service?) and then testing only that business logic by passing in the annotations coming from Genome Nexus.
We can also create integration tests which confirm the proper interactions occur between the genome nexus web server and the genome nexus annotation client ... these would be valuable. But the focus of the integration tests should be to hit every interface dependency (all endpoints) using all valid argument combinations. And the integration tests should not be automatically run during every code build or PR submission, but should be run periodically (probably manually with 'mvn integration-test' or similar)
That's the picture in my mind ... but happy to discuss options.

Reply:
thanks @sheridancbio I think that's a good idea to mock the GN responses